src/promptflow-core/promptflow/core/_prompty_utils.py

```diff
@@ -448,4 +468,165 @@ def parse_chat(chat_str, images: List = None, valid_roles: List[str] = None):
+    else:
+        for i, function in enumerate(functions):
+            validate_function(common_tsg, i, function, ChatAPIInvalidFunctions)
+
+
+def validate_tools(tools):
+    tool_example = json.dumps(
+        {
+            "type": "function",
+            "function": {
+                "name": "get_current_weather",
+                "description": "Get the current weather in a given location",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "location": {
+                            "type": "string",
+                            "description": "The city and state, e.g. San Francisco, CA",
+                        },
+                        "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
+                    },
+                    "required": ["location"],
+                },
+            },
+        }
+    )
+    common_tsg = (
+        f"Here is a valid tool example: {tool_example}. See more details at "
+        "https://platform.openai.com/docs/api-reference/chat/create"
+    )
+
+    if len(tools) == 0:
+        raise ChatAPIInvalidTools(message=f"tools cannot be an empty list. {common_tsg}")
+    for i, tool in enumerate(tools):
+        # validate if the tool is a dict
+        if not isinstance(tool, dict):
+            raise ChatAPIInvalidTools(message=f"tool {i} '{tool}' is not a dict. {common_tsg}")
+        # validate if has required keys
+        for key in ["type", "function"]:
+            if key not in tool.keys():
+                raise ChatAPIInvalidTools(message=f"tool {i} '{tool}' does not have '{key}' property. {common_tsg}")
+        validate_function(common_tsg, i, tool["function"], ChatAPIInvalidTools)
+
+
+def process_function_call(function_call):
+    if function_call is None:
+        param = "auto"
+    elif function_call == "auto" or function_call == "none":
+        param = function_call
+    else:
+        function_call_example = json.dumps({"name": "function_name"})
+        common_tsg = (
+            f"Here is a valid example: {function_call_example}. See the guide at "
+            "https://platform.openai.com/docs/api-reference/chat/create#chat/create-function_call "
+            "or view sample 'How to call functions with chat models' in our gallery."
+        )
+        param = function_call
+        if not isinstance(param, dict):
+            raise ChatAPIInvalidFunctions(
+                message=f"function_call parameter '{param}' must be a dict, but not {type(function_call)}. {common_tsg}"
+            )
+        else:
+            if "name" not in function_call:
+                raise ChatAPIInvalidFunctions(
+                    message=f'function_call parameter {json.dumps(param)} must contain "name" field. {common_tsg}'
+                )
+    return param
+
+
+def process_tool_choice(tool_choice):
```

Comment:
looks like this is just some validate logic

process_tool_choice -> validate_tool_choice

Reply:
rename